File: ruby/binary-search/binary_search.rb
class BinarySearch

  private

  attr_writer :catalog

  def initialize(catalog)
    self.catalog = catalog
  end

  public
  
  attr_reader :catalog
    
  def search_for(item)
    lower = 0
    upper = catalog.size - 1
    while lower <= upper
      mid = (lower + upper) / 2
      return mid if catalog[mid] == item
      catalog[mid] > item ? upper = mid - 1 : lower = mid + 1
    end
  end
    
end

File: python/binary-search/binary_search.py
def find(search_list, value):
    lower = 0
    upper = len(search_list) - 1
    while lower <= upper:
        mid = (lower + upper) // 2
        if search_list[mid] == value:
            return mid
        if search_list[mid] > value:
            upper = mid - 1
        else:
            lower = mid + 1
    raise ValueError("value not in array")
